File: apps/desktop/src/domain/scene/group.ts
// Group/composite world-space resolution (P5-T10) — mirrors
// crates/audiogram-render/src/scene_frame.rs's group-aware `world_transform`/
// `is_visible_with_ancestors`. Depth-capped at 3 (MAX_GROUP_DEPTH) as a
// defensive bound against accidental cycles — the UI itself blocks nesting
// beyond that when grouping (design.slice.ts's groupNodes).
import type { SceneNode, Transform } from '../../types'
import { compose } from './transform'
import { computeEffectiveTransform, isVisibleAt } from './timing'

export const MAX_GROUP_DEPTH = 3

export type NodesById = Map<string, SceneNode>

export function indexById(nodes: SceneNode[]): NodesById {
  return new Map(nodes.map(n => [n.id, n]))
}

/** How many ancestor groups sit above `node` (0 = root level). Used to
 * block grouping beyond `MAX_GROUP_DEPTH` (T10 step 7 edge case). */
export function groupDepth(node: SceneNode, nodesById: NodesById): number {
  let depth = 0
  let current = node
  while (current.parentId && depth <= MAX_GROUP_DEPTH) {
    const parent = nodesById.get(current.parentId)
    if (!parent) break
    depth += 1
    current = parent
  }
  return depth
}

/** True if `node` and every ancestor group up the `parentId` chain is
 * visible at `t` — a group's `timing` hides the whole subtree (T10 step 6),
 * same mechanism as a leaf node's own `timing`, just walked upward. */
export function isVisibleWithAncestors(node: SceneNode, nodesById: NodesById, t: number, dur: number, depth = 0): boolean {
  if (!isVisibleAt(node, t, dur)) return false
  if (node.parentId && depth < MAX_GROUP_DEPTH) {
    const parent = nodesById.get(node.parentId)
    if (parent) return isVisibleWithAncestors(parent, nodesById, t, dur, depth + 1)
  }
  return true
}

/** World-space (root canvas) transform for `node` — composes its own
 * (possibly time-animated) transform with every ancestor group's own
 * (possibly time-animated) transform, parent ⊗ child up the chain. A
 * group's animIn/animOut therefore applies to the whole subtree "for
 * free": compose() multiplies opacity and offsets position, so a fading
 * group fades every descendant along with it. */
export function worldTransform(node: SceneNode, nodesById: NodesById, t: number, dur: number, depth = 0): Transform {
  const own = computeEffectiveTransform(node, t, dur)
  if (node.parentId && depth < MAX_GROUP_DEPTH) {
    const parent = nodesById.get(node.parentId)
    if (parent) return compose(worldTransform(parent, nodesById, t, dur, depth + 1), own)
  }
  return own
}

/** Same as `worldTransform`, but ignores timing/animIn/animOut — for
 * structural edits (group/ungroup, T10) where "current position" means the
 * node's own static transform, not a snapshot of a possibly-mid-animation
 * runtime state. */
export function staticWorldTransform(node: SceneNode, nodesById: NodesById, depth = 0): Transform {
  if (node.parentId && depth < MAX_GROUP_DEPTH) {
    const parent = nodesById.get(node.parentId)
    if (parent) return compose(staticWorldTransform(parent, nodesById, depth + 1), node.transform)
  }
  return node.transform
}

File: crates/audiogram-render/src/group.rs
//! Group/composite world-space resolution (P5-T10) — mirrors
//! `apps/desktop/src/domain/scene/group.ts` exactly. Depth-capped at
//! `MAX_GROUP_DEPTH` (3) as a defensive bound against accidental cycles.
use std::collections::HashMap;

use crate::timing::{effective_transform, is_visible_at};
use audiogram_core::entities::scene_node::Transform;
use audiogram_core::entities::SceneNode;

pub const MAX_GROUP_DEPTH: usize = 3;

pub type NodesById<'a> = HashMap<&'a str, &'a SceneNode>;

pub fn index_by_id(nodes: &[SceneNode]) -> NodesById<'_> {
    nodes.iter().map(|n| (n.id.as_str(), n)).collect()
}

/// True if `node` and every ancestor group up the `parent_id` chain is
/// visible at `t_sec` — a group's `timing` hides the whole subtree.
pub fn is_visible_with_ancestors(node: &SceneNode, nodes_by_id: &NodesById, t_sec: f64, dur: f64) -> bool {
    is_visible_with_ancestors_depth(node, nodes_by_id, t_sec, dur, 0)
}

fn is_visible_with_ancestors_depth(node: &SceneNode, nodes_by_id: &NodesById, t_sec: f64, dur: f64, depth: usize) -> bool {
    if !is_visible_at(node, t_sec, dur) {
        return false;
    }
    if let Some(parent_id) = &node.parent_id {
        if depth < MAX_GROUP_DEPTH {
            if let Some(parent) = nodes_by_id.get(parent_id.as_str()) {
                return is_visible_with_ancestors_depth(parent, nodes_by_id, t_sec, dur, depth + 1);
            }
        }
    }
    true
}

/// World-space transform for `node` — composes its own (possibly time-
/// animated) transform with every ancestor group's own transform, parent ⊗
/// child up the chain (Transform::compose, tested at T2).
pub fn world_transform(node: &SceneNode, nodes_by_id: &NodesById, t_sec: f64, dur: f64) -> Transform {
    world_transform_depth(node, nodes_by_id, t_sec, dur, 0)
}

fn world_transform_depth(node: &SceneNode, nodes_by_id: &NodesById, t_sec: f64, dur: f64, depth: usize) -> Transform {
    let own = effective_transform(node, t_sec, dur);
    if let Some(parent_id) = &node.parent_id {
        if depth < MAX_GROUP_DEPTH {
            if let Some(parent) = nodes_by_id.get(parent_id.as_str()) {
                let parent_world = world_transform_depth(parent, nodes_by_id, t_sec, dur, depth + 1);
                return Transform::compose(&parent_world, &own);
            }
        }
    }
    own
}

#[cfg(test)]
mod tests {
    use super::*;
    use audiogram_core::entities::scene_node::{AnimationClip, AnimationId, SceneNodeType, Timing};

    fn make_transform(x: f32, y: f32, w: f32, h: f32) -> Transform {
        Transform { x, y, w, h, rotation: 0.0, opacity: 1.0 }
    }

    fn group_node(id: &str, transform: Transform) -> SceneNode {
        SceneNode {
            id: id.into(), r#type: SceneNodeType::Group, parent_id: None,
            transform, z: 0, timing: None, anim_in: None, anim_out: None, keyframes: vec![], props: None,
        }
    }

    fn child_node(id: &str, parent_id: &str, transform: Transform) -> SceneNode {
        SceneNode {
            id: id.into(), r#type: SceneNodeType::Waveform, parent_id: Some(parent_id.into()),
            transform, z: 0, timing: None, anim_in: None, anim_out: None, keyframes: vec![], props: None,
        }
    }

    #[test]
    fn composes_parent_and_child() {
        let group = group_node("g", make_transform(0.2, 0.2, 0.4, 0.4));
        let child = child_node("c", "g", make_transform(0.1, 0.1, 0.5, 0.5));
        let nodes = [group, child.clone()];
        let by_id = index_by_id(&nodes);
        let world = world_transform(&child, &by_id, 0.0, 0.0);
        let eps = 1e-6_f32;
        assert!((world.x - 0.24).abs() < eps);
        assert!((world.y - 0.24).abs() < eps);
        assert!((world.w - 0.20).abs() < eps);
        assert!((world.h - 0.20).abs() < eps);
    }

    #[test]
    fn group_timing_hides_children_with_no_timing_of_their_own() {
        let mut group = group_node("g", make_transform(0.0, 0.0, 1.0, 1.0));
        group.timing = Some(Timing { start: 2.0, end: 5.0 });
        let child = child_node("c", "g", make_transform(0.0, 0.0, 1.0, 1.0));
        let nodes = [group, child.clone()];
        let by_id = index_by_id(&nodes);
        assert!(!is_visible_with_ancestors(&child, &by_id, 1.0, 0.0));
        assert!(is_visible_with_ancestors(&child, &by_id, 3.0, 0.0));
        assert!(!is_visible_with_ancestors(&child, &by_id, 6.0, 0.0));
    }

    #[test]
    fn group_anim_in_scales_child_opacity() {
        let mut group = group_node("g", make_transform(0.1, 0.1, 0.5, 0.5));
        group.timing = Some(Timing { start: 0.0, end: 10.0 });
        group.anim_in = Some(AnimationClip { preset: AnimationId::Fade, duration: 2.0 });
        let child = child_node("c", "g", make_transform(0.0, 0.0, 1.0, 1.0));
        let nodes = [group, child.clone()];
        let by_id = index_by_id(&nodes);
        let at_start = world_transform(&child, &by_id, 0.0, 0.0);
        assert!((at_start.opacity - 0.0).abs() < 1e-6);
        let at_end = world_transform(&child, &by_id, 2.0, 0.0);
        assert!((at_end.opacity - 1.0).abs() < 1e-6);
    }
}
